## The POP (Pregancny Outcome Prediction) Study
> The POPs study included 4,512 nulliparous women with a viable singleton pregnancy attending The [Rosie Hospital](https://www.cuh.nhs.uk/rosie-hospital/) (Cambridge, UK) for their dating ultrasound scan between 2008 and 2012 ([Pasupathy et al 2008](https://bmcpregnancychildbirth.biomedcentral.com/articles/10.1186/1471-2393-8-51), [Gaccioli et al 2016](https://www.sciencedirect.com/science/article/pii/S0143400416305586)). Women were serially scanned through the pregnancy and had blood obtained at recruitment, 20, 28 and 36 weeks gestational age. DNA samples and height/weight measurements of the partners were also collected. After delivery, biopsies of the placenta, placental membranes, umbilical cord, and cord blood were collected. Thorough design of POPs and careful sample collection allowed us to create an extensive, optimally phenotyped biobank of complicated pregnancies and controls, including approximately 230,000 blood and tissue samples stored in -80°C freezers, 24,000 formalin fixed placental biopsies, and 4,000 paternal DNA samples. Read more [here](https://www.obgyn.cam.ac.uk/).

## Placenta Samples
> All the samples were obtained from the Pregnancy Outcome Prediction Study (**[POPS](https://www.obgyn.cam.ac.uk/research/pops-2/)**) – a prospective cohort study of nulliparous women attending the Rosie Hospital, Cambridge (UK) for their dating ultrasound scan between January 14, 2008, and July 31, 2012.

> A total of __302__ unique placenta samples were considered in this study, among which there were 94 PE, 56 FGR, and 155 control samples – three cases of samples were classified as both PE and FGR. 

> We generated __324__ total RNA-Seq data sets from __302__ unique placental biopsies – __22__ biopsies were sequenced twice (__19__ control samples and __3__ samples classified as both PE and FGR cases). 

> Out of the __302__ placental biopsies, __three__ samples were excluded due to the presence of decidual contamination of the placental biopsies.

> __Three__ cases (i.e. samples classified as both PE and FGR) were also excluded in the analyses (except differentially expressed gene analysis). 

> A further sample was flagged as failed at the library preparation process.

> Thus, a total of __295__ unique placental biopsies were used (__159__ male and __136__ female placentas).

## Annotations Used 
> GRCh38.p3

> [Ensembl 82](ftp://ftp.ensembl.org/pub/release-82/gtf/homo_sapiens)

> [miRBase v21](ftp://mirbase.org/pub/mirbase/21/)

> [piRBase v1.0](http://www.regulatoryrna.org/database/piRNA/download/archive/v1.0/)

## Acknowledgement
> This project is fully funded by the [NIHR Cambridge Biomedical Research Centre](https://cambridgebrc.nihr.ac.uk/)

## Disclaimer
> The views expressed are those of the author(s) and not necessarily those of the NHS, the NIHR or the Department of Health and Social Care.

## Reference
> Under review...

## Availability
> The source code is available [here](https://github.com/sung/ShinyPlacentome)

> RNA-seq data have been deposited in the [European genome-phenome archive](https://ega-archive.org/), with an accession number [EGAS00001002205](https://ega-archive.org/studies/EGAS00001002205).

***
Developed and maintained by [Sung Gong](https://www.obgyn.cam.ac.uk/staff/research-staff/sung-gong/)


[![University of Cambridge](www/images/Uni-Camb-Medic-Logo.png)](https://www.cam.ac.uk)
[<img src="www/images/Cambridge_BRC_NIHR_logo_2019_RGB.jpeg" width=350>](https://cambridgebrc.nihr.ac.uk/)
